stl/inc/ranges

```diff
@@ -2260,6 +2214,416 @@ namespace ranges {
+        };
+        // clang-format on
+
+        template <bool _Const>
+        class _Iterator : public _Category_base<iterator_traits<iterator_t<_Maybe_const<_Const, _Vw>>>,
+                              iterator_traits<iterator_t<range_reference_t<_Maybe_const<_Const, _Vw>>>>,
+                              is_reference_v<range_reference_t<_Maybe_const<_Const, _Vw>>>> {
+        private:
+            template <bool>
+            friend class _Iterator;
+            template <bool>
+            friend class _Sentinel;
+
+            using _Parent_t  = _Maybe_const<_Const, join_view>;
+            using _Base      = _Maybe_const<_Const, _Vw>;
+            using _SubRng    = range_reference_t<_Base>;
+            using _OuterIter = iterator_t<_Base>;
+            using _InnerIter = iterator_t<_SubRng>;
+
+            // True if and only if the expression *i, where i is an iterator from the outer range, is a glvalue:
+            static constexpr bool _Deref_is_glvalue = is_reference_v<_SubRng>;
+
+            /* [[no_unique_address]] */ _OuterIter _Outer{};
+            /* [[no_unique_address]] */ _InnerIter _Inner{};
+            _Parent_t* _Parent{};
+
+            constexpr void _Satisfy() {
+                auto _Update_inner = [this](_SubRng _Range) -> auto& {
+                    if constexpr (_Deref_is_glvalue) {
+                        (void) this; // Suppress Clang "unused capture" warning
+                        return _Range;
+                    } else {
+                        return (_Parent->_Inner = views::all(_STD move(_Range)));
+                    }
+                };
+
+                for (; _Outer != _RANGES end(_Parent->_Range); ++_Outer) {
+                    auto& _Tmp = _Update_inner(*_Outer);
+                    _Inner     = _RANGES begin(_Tmp);
+                    if (_Inner != _RANGES end(_Tmp)) {
+                        return;
+                    }
+                }
+                if constexpr (_Deref_is_glvalue) {
+                    _Inner = _InnerIter{};
+                }
+            }
+
+#if _ITERATOR_DEBUG_LEVEL != 0
+            constexpr void _Check_dereference() const noexcept {
+                _STL_VERIFY(_Parent != nullptr, "cannot dereference value-initialized join_view iterator");
+                _STL_VERIFY(_Outer != _RANGES end(_Parent->_Range), "cannot dereference end join_view iterator");
+                sentinel_t<_InnerRng> _Last;
+                if constexpr (_Deref_is_glvalue) {
+                    _Last = _RANGES end(*_Outer);
+                } else {
+                    _Last = _RANGES end(_Parent->_Inner);
+                }
+                _STL_VERIFY(_Inner != _Last, "cannot dereference end join_view iterator");
+            }
+
+            constexpr void _Same_range(const _Iterator& _Right) const noexcept {
+                _STL_VERIFY(_Parent == _Right._Parent, "cannot compare incompatible join_view iterators");
+            }

```

Comment:
Should we inline this?

Reply:
I have a mild preference to avoid repetition. Notably we didn't inline `_Same_range` in `transform_view::_Iterator`.